### PR TITLE
Adds ability to query payments by payment id

### PR DESCRIPTION
**Pull request by clarissalimab**
Closes [tix #3079](https://github.com/CommitChange/tix/issues/3079).

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -113,7 +113,7 @@
     resource(:miscellaneous_np_info, {only: [:show, :update]})
 
 		namespace(:button) do
-			root({to: :advanced})
+			root({action: :advanced})
```
Comment (clarissalimab) [inline config/routes.rb]: I did this tiny change because this deprecation warning was very annoying and it was easy to resolve, but I can do it on other PR if you prefer :smile_cat:

File: lib/search_vector.rb [hunk 2]
```diff
@@ -7,8 +7,9 @@ module SearchVector
   def self.query(query_string, expr=nil)
     (expr || Qexpr.new).where(
       "(supporters.fts @@ websearch_to_tsquery('english', $search)
-      OR donations.fts  @@ websearch_to_tsquery('english', $search))",
-      { search: query_string }
+      OR donations.fts  @@ websearch_to_tsquery('english', $search)
+      OR payments.id::varchar(255) = $search)",
```
Comment (wwahammy) [inline lib/search_vector.rb]: Hmmm, this worries me. By converting to varchar, it will likely bypass the index on the id column and then lead to running a search that has to go through every possible payment. Can you run an a query analysis of this query, i.e. call the query in the dbconsole but add `EXPLAIN ANALYZE` at the beginning?

File: lib/search_vector.rb [hunk 3]
```diff
@@ -7,8 +7,9 @@ module SearchVector
...
+      { search: query_string.to_s }
```
Comment (wwahammy) [inline lib/search_vector.rb]: Why the change here?

**Review by wwahammy — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 1 above)
Comment (wwahammy) [reply config/routes.rb]: Another PR would be ideal.

File: lib/search_vector.rb (hunk 3 above)
Comment (clarissalimab) [reply lib/search_vector.rb]: What change do you mean?

Comment (wwahammy) [reply lib/search_vector.rb]: You added a `.to_s` to `search_query`

File: lib/search_vector.rb (hunk 2 above)
Comment (clarissalimab) [reply lib/search_vector.rb]: It became really slow, I didn't realize that. I did a workaround to prevent that.

File: lib/search_vector.rb (hunk 3 above)
Comment (clarissalimab) [reply lib/search_vector.rb]: I had a problem passing an integer to `websearch_to_tsquery`

File: lib/core_ext/string.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# License: AGPL-3.0-or-later WITH Web-Template-Output-Additional-Permission-3.0-or-later
+
+class String
+  def is_int?
+    !!(self =~ /\A\d+\z/)
+  end
+end
```
Comment (wwahammy) [inline lib/core_ext/string.rb]: Why this design decision to create this core_exts folder? I don't think it's bad, in fact, I can see advantages to doing it this way. How did the decision come about?

File: lib/search_vector.rb [hunk 5]
```diff
@@ -5,10 +5,13 @@ module SearchVector
   AcceptedTables = ['supporters', 'payments']
 
   def self.query(query_string, expr=nil)
-    (expr || Qexpr.new).where(
+    query = if query_string.is_a?(Integer) || query_string.is_int?
+      "payments.id = $search"
+    else
       "(supporters.fts @@ websearch_to_tsquery('english', $search)
-      OR donations.fts  @@ websearch_to_tsquery('english', $search))",
-      { search: query_string }
```
Comment (wwahammy) [inline lib/search_vector.rb]: When an integer is passing, I think we should still be searching in the `supporter.fts` or `donations.fts` in addition to the `payments.id`.

**Review by wwahammy — CHANGES_REQUESTED** [review]

File: lib/core_ext/string.rb (hunk 4 above)
Comment (clarissalimab) [reply lib/core_ext/string.rb]: I did some research on how to do it, I saw that some people created just an `ext` folder and someone suggested creating as `core_ext`. I checked how it is being done on Rails and that's the name of the folder that holds each ActiveSupport type classes: https://github.com/rails/rails/tree/main/activesupport/lib/active_support/core_ext

I could also follow the same pattern and create a folder called `string` and a file called `string.rb`, and a class called `types.rb` that checks if the string is compatible with some type. For example:

```
├── lib
│   ├── core_ext
│   │   ├── string
│   │   │   ├── types.rb
│   │   ├── string.rb
```

File: lib/search_vector.rb [hunk 6]
```diff
@@ -5,10 +5,15 @@ module SearchVector
   AcceptedTables = ['supporters', 'payments']
 
   def self.query(query_string, expr=nil)
-    (expr || Qexpr.new).where(
+    query = if query_string.is_a?(Integer) || query_string.is_int?
+      "(supporters.fts @@ websearch_to_tsquery('english', $search::varchar(255))
```
Comment (wwahammy) [inline lib/search_vector.rb]: When you didn't cast to a varchar, did it cause problems here?

Comment (clarissalimab) [reply lib/search_vector.rb]: Yes, if I pass an integer I get an error:

```
ActiveRecord::StatementInvalid:
       PG::UndefinedFunction: ERROR:  function websearch_to_tsquery(unknown, integer) does not exist
```

Although when it comes from the interface, the query already comes as a string, I think it won't hurt to avoid problems.

Comment (wwahammy) [reply lib/search_vector.rb]: Hmmm, that makes sense then.

File: spec/lib/query/query_payments_spec.rb [hunk 7]
```diff
@@ -294,6 +295,27 @@ def generate_donation(h)
...
+
+        it 'returns one result' do
+          InsertDonation.with_stripe(input)
+          donation_result_tomorrow
+          donation_result_yesterday
+
+          result = QueryPayments::full_search(nonprofit.id, { search: Payment.last.id })
+          expect(result[:data].count).to eq 1
+        end
+      end
```
Comment (wwahammy) [inline spec/lib/query/query_payments_spec.rb]: Could you add a test that makes sure it still finds a payments if an integer is in the notes but isn't the payment ID itself?

File: lib/search_vector.rb [hunk 8]
```diff
@@ -5,10 +5,15 @@ module SearchVector
...
+      OR donations.fts  @@ websearch_to_tsquery('english', $search::varchar(255))
```
Comment (wwahammy) [inline lib/search_vector.rb]: Is there any reason we couldn't move these two out lines out since they're the same between each branch of the boolean and then only append the `OR payments.id = $search::INTEGER` to the statement if the boolean is true?

**Review by wwahammy — CHANGES_REQUESTED** [review]
Looking pretty good, just a few tweaks and it'll be ready to go!

File: lib/core_ext/string.rb (hunk 4 above)
Comment (wwahammy) [reply lib/core_ext/string.rb]: I think the design you have now makes sense. Nice job!

File: spec/lib/query/query_payments_spec.rb (hunk 7 above)
Comment (clarissalimab) [reply spec/lib/query/query_payments_spec.rb]: I added one test for this :smiley:

File: lib/search_vector.rb (hunk 8 above)
Comment (clarissalimab) [reply lib/search_vector.rb]: I thought it looked better this way, but I don't know. I did it as you suggested now, what do you think?

Comment (wwahammy) [reply lib/search_vector.rb]: Haha, I looked at it now and I agree with you, it looked better before. Sorry!

Comment (clarissalimab) [reply lib/search_vector.rb]: Hahaha I'll change it back

**Review by wwahammy — APPROVED** [review]
Fantastic work as always @clarissalimab!
